api/v1alpha1/ratelimit_types.go

```diff
@@ -91,6 +92,80 @@ type RateLimitRule struct {
 	// 429 HTTP status code is sent back to the client when
 	// the selected requests have reached the limit.
 	Limit RateLimitValue `json:"limit"`
+	// CostPerRequest specifies the number to reduce the rate limit counters
+	// on the request path. If this is not specified, the default behavior
+	// is to reduce the rate limit counters by 1.
+	//
+	// When Envoy receives a request that matches the rule, it tries to reduce the
+	// rate limit counters by the specified number. If the counter doesn't have
+	// enough capacity, the request is rate limited.
+	//
+	// +optional
+	// +notImplementedHide
+	CostPerRequest *RateLimitCost `json:"costPerRequest,omitempty"`
+	// CostPerResponse specifies the number to reduce the rate limit counters
+	// after the response is sent back to the client or the request stream is closed.
+	//
+	// The cost is used to reduce the rate limit counters for the matching requests.
+	// Since the reduction happens after the request stream is complete, the rate limit
+	// won't be enforced for the current request, but for the subsequent matching requests.
+	//
+	// This is optional and if not specified, the rate limit counters are not reduced
+	// on the response path.
+	//
+	// Currently, this is only supported for HTTP Global Rate Limits.
+	//
+	// +optional
+	// +notImplementedHide
+	CostPerResponse *RateLimitCost `json:"costPerResponse,omitempty"`
+}
+
+// RateLimitCost specifies where the Envoy retrieves the number to reduce the rate limit counters.
+//
+// +kubebuilder:validation:XValidation:rule="!(has(self.number) && has(self.dynamicMetadata))",message="only one of number or dynamicMetadata can be specified"
+type RateLimitCost struct {
+	// Type specifies the source of the rate limit cost. Currently, only "Number" and "DynamicMetadata" are supported.
+	//
+	// +kubebuilder:validation:Required
+	Type RateLimitCostType `json:"type"`
+	// Number specifies the fixed usage number to reduce the rate limit counters.
+	// Using zero can be used to only check the rate limit counters without reducing them.
+	//
+	// +optional
+	// +notImplementedHide
+	Number *uint64 `json:"number,omitempty"`
+	// DynamicMetadata specifies the filter metadata to retrieve the usage number from.
+	//
+	// +optional
+	// +notImplementedHide
+	DynamicMetadata *RateLimitCostDynamicMetadata `json:"dynamicMetadata,omitempty"`
+}
+
+// RateLimitCostType specifies the source of the rate limit cost.
+// Valid RateLimitCostType values are "Number" and "DynamicMetadata".
+//
+// +kubebuilder:validation:Enum=Number;DynamicMetadata
+type RateLimitCostType string
+
+const (
+	// RateLimitCostTypeNumber specifies the rate limit cost to be a fixed number.
+	RateLimitCostTypeNumber RateLimitCostType = "Number"
```

Comment:
can we leave out `Number` for now from the API ? is there a use case for it ?
for e.g. if I want to ratelimit at 100 counts/sec, and each requests costs me 2, I can easily update the limit to 50 counts/sec

Reply:
oh is number used here, so we can set request rate limit cost to `0` ?

Reply:
yes

Reply:
I explicitly added that comment about 0 can be used to do "just checking"